controlplane/controllers/orchestrated_inplace_upgrade_controller.go

```diff
@@ -0,0 +1,301 @@
+type OrchestratedInPlaceUpgradeScope struct {
+	cluster          *clusterv1.Cluster
+	ck8sControlPlane *controlplanev1.CK8sControlPlane
+	ck8sPatcher      inplace.Patcher
+	upgradeTo        string
+	ownedMachines    collections.Machines
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *OrchestratedInPlaceUpgradeController) SetupWithManager(mgr ctrl.Manager) error {
+	r.scheme = mgr.GetScheme()
+	r.recorder = mgr.GetEventRecorderFor("ck8s-cp-orchestrated-inplace-upgrade-controller")
+	r.machineGetter = &ck8s.Management{
+		Client: r.Client,
+	}
+	r.lock = inplace.NewUpgradeLock(r.Client)
+
+	if err := ctrl.NewControllerManagedBy(mgr).
+		For(&controlplanev1.CK8sControlPlane{}).
+		Owns(&clusterv1.Machine{}).
+		Complete(r); err != nil {
+		return fmt.Errorf("failed to get new controller builder: %w", err)
+	}
+
+	return nil
+}
+
+// +kubebuilder:rbac:groups=core,resources=configmaps,verbs=get;create;delete
+// +kubebuilder:rbac:groups=cluster.x-k8s.io,resources=clusters;clusters/status,verbs=get;list;watch
+// +kubebuilder:rbac:groups=cluster.x-k8s.io,resources=machines;machines/status,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=cluster.x-k8s.io,resources=machinesets;machinesets/status,verbs=get;list;watch
+// +kubebuilder:rbac:groups=apiextensions.k8s.io,resources=customresourcedefinitions,verbs=get;list;watch
+
+// Reconcile handles the reconciliation of a CK8sControlPlane object.
+func (r *OrchestratedInPlaceUpgradeController) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	traceID := trace.NewID()
+	log := r.Log.WithValues("orchestrated_inplace_upgrade", req.NamespacedName, "trace_id", traceID)
+	log.V(1).Info("Reconciliation started...")
+
+	ck8sCP := &controlplanev1.CK8sControlPlane{}
+	if err := r.Get(ctx, req.NamespacedName, ck8sCP); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.V(1).Info("CK8sControlPlane resource not found. Ignoring since the object must be deleted.")
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, fmt.Errorf("failed to get CK8sControlPlane: %w", err)
+	}
+
+	if inplace.GetUpgradeInstructions(ck8sCP) == "" {
+		log.V(1).Info("CK8sControlPlane has no upgrade instructions, skipping reconciliation")
+		return ctrl.Result{}, nil
+	}
+
+	if isDeleted(ck8sCP) {
+		log.V(1).Info("CK8sControlPlane is being deleted, skipping reconciliation")
+		return ctrl.Result{}, nil
+	}
+
+	if !r.machinesAreReady(ck8sCP) {
+		log.V(1).Info("Machines are not ready, requeuing...")
+		return ctrl.Result{RequeueAfter: 5 * time.Second}, nil
+	}
+
+	scope, err := r.createScope(ctx, ck8sCP)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to create scope: %w", err)
+	}
+
+	lockedMachine, err := r.lock.IsLocked(ctx, scope.cluster)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to check if upgrade is locked: %w", err)
+	}
+
+	// Upgrade is locked and a machine is already upgrading
+	if lockedMachine != nil {
```

Comment:
What happens on if upgrade is failed on the machine, do we wait/hold the lock until it succeds?

Reply:
Thanks for pointing this out, I noticed that I had to move the `IsMachineUpgradeFailed` check a bit higher.

In short, yes. Lemme explain a bit further:

### Scenario 1: Not locked
- Let's say the upgrade process is not locked. We select a machine to upgrade (let's call it `M`) and lock the upgrade process while mentioning in the lock object that machine `M` is upgrading.

### Scenario 2: Locked and the machine is upgraded
- We see that the process is locked, so we get machine `M` and notice that it is upgraded (`release` annotation). We unlock and requeue the reconciliation.

### Scenario 3: Locked and the machine is not upgraded (still upgrading)
- We requeue the reconciliation.

### Scenario 4: A machine that we tried to upgrade before is not there any more (somehow deleted)
- The lock config map has the machine information and tries to obtain it but notices that it's not there, so the `IsLocked` won't return an `upgradingMachine`, the lock config map is deleted (`Unlock` is called) and we continue on to select a fresh machine to upgrade.

### Scenario 5: Locked and the machine does not have any upgrade instructions
- This is a weird situation (maybe the admin deleted the annotations accidentally), but _can_ happen nonetheless. We consider that a stale lock and Unlock and requeue.

### Scenario 6: Locked and the machine upgrade failed
- We mark the `CK8sControlPlane` object with `failed` status and publish the respective event and retry the process (basically the underlying `single machine in-place upgrade reconciler` is doing the retry and we just wait).

Reply:
Thanks for the list - this was very helpful for the review.